### PR TITLE
[llvm][CodeGen] Resolve issues when updating live intervals in window scheduler

### PR DESCRIPTION
**Pull request by huaatian**
Corrupted live interval information can cause window scheduling to crash in some cases. 
By adding the missing MBB's live interval information in the ModuloScheduleExpander, 
the information can be correctly analyzed in the window scheduler.

### CODE REVIEW

**Comment by huaatian** [comment]
This bug was reported in https://github.com/llvm/llvm-project/pull/99454.
The root cause of the issue is that slot index and live interval information are required when constructing TripleMBB. During the analysis of some virtual registers, references are made to an MBB that has already been processed by the machine pipeliner. Since the machine pipeliner does not preserve live interval and slot index information, the following error occurs:
image
The probability of this situation occurring is low. Typically, when updating the LiveIntervals in TripleMBB, scheduled MBBs are not referenced. Therefore, in such cases, we reanalyze the live intervals.
PS :This error scenario is difficult to reproduce, so we can only try to simplify the reported scenario as much as possible.

**Comment by huaatian** [comment]
Hi, could you please take some time to review this PR? Sorry for taking up your time, and thank you! @arsenm

**Comment by huaatian** [comment]
ping

**Comment by huaatian** [comment]
hi，@arsenm，could you please take a look at this PR？Thanks for your time!

File: llvm/lib/CodeGen/WindowScheduler.cpp [hunk 1]
```diff
@@ -644,6 +645,20 @@ void WindowScheduler::expand() {
...
+void WindowScheduler::reanalyzeLiveIntervals() {
+  auto SlotIndex = Context->LIS->getSlotIndexes();
+  // Check if the SlotIndex infomation is missing.
+  for (auto &MBB : *MF)
+    for (auto &MI : MBB)
+      if (!SlotIndex->hasIndex(MI)) {
+        // The slot index and live intervals of MF have been corrupted and need
+        // to be reanalyzed.
+        SlotIndex->reanalyze(*MF);
+        Context->LIS->reanalyze(*MF);
```
Comment (arsenm) [inline llvm/lib/CodeGen/WindowScheduler.cpp:657]: This is much too heavy of a hammer. You should be updating specific intervals as you change them (probably need to add some handleMove calls somewhere)

File: llvm/test/CodeGen/Hexagon/swp-ws-live-intervals.mir [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+# RUN: llc --march=hexagon %s -run-pass=pipeliner -debug-only=pipeliner \
+# RUN: -window-sched=force -filetype=null -window-search-num=100 \
+# RUN: -window-search-ratio=100 -window-diff-limit=0 -verify-machineinstrs \
+# RUN: 2>&1 | FileCheck %s
+
+# The bug was reported at https://github.com/llvm/llvm-project/pull/99454. 
+# It is caused by the corruption of live intervals in certain scenarios.
+#
+# CHECK: Best window offset is {{[0-9]+}} and Best II is {{[0-9]+}}.
+# CHECK: Best window offset is {{[0-9]+}} and Best II is {{[0-9]+}}.
```
Comment (arsenm) [inline llvm/test/CodeGen/Hexagon/swp-ws-live-intervals.mir:12]: Better check would check the actual scheduling result

Comment (huaatian) [reply llvm/test/CodeGen/Hexagon/swp-ws-live-intervals.mir]: Updated

File: llvm/lib/CodeGen/WindowScheduler.cpp (hunk 1 above)
Comment (huaatian) [reply llvm/lib/CodeGen/WindowScheduler.cpp]: You are right. 
The issue occurs in the ModuloScheduleExpander module. We have avoided the current problem by adding the minimal amount of live interval information, without affecting the window scheduler.

**Review by arsenm — APPROVED** [review]
I'm surprised this is enough to update everything

**Comment by huaatian** [comment]
> I'm surprised this is enough to update everything

This patch only adds the MBB information to the maps. 
In the window scheduler, we have already call the repairIntervalsInRange() interface to fix the missing live interval information. 
The issue occurred because the live interval information corresponding to the newly generated code after pipeliner scheduling was corrupted. 
By simply adding the MBB information, the repairIntervalsInRange() interface will no longer fail.

**Comment by huaatian** [comment]
/cherrypick https://github.com/llvm/llvm-project/commit/43ba1097ee747b4ec5e757762ed0c9df6255a292

**Comment by huaatian** [comment]
/cherry-pick https://github.com/llvm/llvm-project/commit/43ba1097ee747b4ec5e757762ed0c9df6255a292

**Pull request merged**